---
title: "Prevalence of cysteine degrading bacteria in the healthy human gut"
author: "Domenick J. Braccia"
date: "`r format(Sys.time(), '%d %B, %Y')`"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Overview

TODOs:
* [X] - generate prevalence of Sulfate and Sulfite Reducing Bacteria
* [ ] - report above prevalence better.. in a table or something

* [X] - run GutFunFind on known genes and then expand list of cysteine reducing bacteria 
* [X] - import list of cysteine degrading bacteria to this script
* [X] - run same set of analyses on these names as the sulfate / sulfite reducing bac.

* [ ] - make boxplots for RA of sulfi(a)te red. bac. and RA of cys. deg. bac.


# Loading Libraries

```{r, message=FALSE, warning=FALSE}
library(curatedMetagenomicData, quietly = TRUE)
library(tictoc)
library(dplyr)
library(ggplot2)
```


# Loading data

```{r}
if(file.exists("../data/all_data.RDS")) {
  all_data <- readRDS("../data/all_data.RDS")
} else {
  tic()
  all_data <- curatedMetagenomicData("*metaphlan_bugs_list.stool*", dryrun = FALSE)
  saveRDS(all_data, file = "../data/all_data.RDS")
  toc()
}
```

# Exploring data

```{r}
## merging all datasets
all_data_merged <- mergeData(all_data)

all_pData <- pData(all_data_merged)
disease_type <- all_pData %>% 
  select("study_condition", "disease", "disease_subtype", "disease_stage")
sum(is.na(all_pData$disease))
sum(all_pData$disease == "healthy")
study_cond <- all_pData$study_condition
study_cond_NA <- all_data_merged[,is.na(all_pData$study_condition) ]

```

# Processing and subsetting data

```{r}
## removing NAs from pData
all_pData <- all_pData[!is.na(all_pData$study_condition), ]

## subsetting pData to study_condition == "controls"
all_pData_controls <- all_pData %>% filter(study_condition == "control")

## getting rid of NAs in relative abundance (RA) data
all_RA <- exprs(all_data_merged)
all_RA <- all_RA[, !is.na(all_pData$study_condition)]

## subsetting RA down to just present in control samples
control_RA <- all_RA[, colnames(all_RA) %in% rownames(all_pData_controls)]
```

# Prevalence of sulfite reducing bacteria in healthy controls

```{r}
# getting B. wadsworthia prevalence
b_wad <- control_RA[grep("s__Bilophila_wadsworthia$", rownames( control_RA)), ]
b_wad_summary <- summary(b_wad)
b_wad_sd <- sd(b_wad)
b_wad_prev <- sum(b_wad > 0) / length(b_wad)
```

Across all healthy controls available in `curatedMetagenomicData` (52 datasets), the sulfite reducing bacteria Bilophila wadsworthia was detected in `r round(b_wad_prev*100, 1)`% of samples at an average relative abundance of `r b_wad_summary[4]`% (median relative abundance = `r b_wad_summary[3]`%).

# Prevalence of sulfate reducing bacteria

```{r}
## getting Desulfovibrio spp: piger, desulfuricans, gigas
d_piger <- control_RA[grep("s__Desulfovibrio_piger$", rownames(control_RA)), ]
d_piger_summary <- summary(d_piger)
d_piger_sd <- sd(d_piger)
d_piger_prev <- sum(d_piger > 0) / length(d_piger)

## desulfricans
d_desulf <- control_RA[grep("s__Desulfovibrio_desulfuricans$", rownames(control_RA)), ]
d_desulf_summary <- summary(d_desulf)
d_desulf_sd <- sd(d_desulf)
d_desulf_prev <- sum(d_desulf > 0) / length(d_desulf)

## gigas
d_gigas <- control_RA[grep("s__Desulfovibrio_gigas$", rownames(control_RA)), ]
d_gigas_summary <- summary(d_gigas)
d_gigas_sd <- sd(d_gigas)
d_gigas_prev <- sum(d_gigas > 0) / length(d_gigas)

## presence of one or more of above 4 species
sulf_red_RA <- data.frame(b_wad = b_wad, 
                       d_piger = d_piger,
                       d_desulf = d_desulf,
                       d_gigas = d_gigas)
sulf_red_RA_rs <- rowSums(sulf_red_RA)
sulf_red_RA_prev <- sum(sulf_red_RA_rs > 0) / length(sulf_red_RA_rs)
```

Across all healthy controls available in `curatedMetagenomicData` (52 datasets), the sulfate reducing bacteria Desulfovibrio piger was detected in `r round(d_piger_prev*100, 1)`% of samples at an average relative abundance of `r d_piger_summary[4]`% (median relative abundance = `r d_piger_summary[3]`%).

Across all healthy controls available in `curatedMetagenomicData` (52 datasets), the sulfate reducing bacteria Desulfovibrio desulfuricans was detected in `r round(d_desulf_prev*100, 1)`% of samples at an average relative abundance of `r d_desulf_summary[4]`% (median relative abundance = `r d_desulf_summary[3]`%).

Across all healthy controls available in `curatedMetagenomicData` (52 datasets), the sulfate reducing bacteria Desulfovibrio gigas was detected in `r round(d_gigas_prev*100, 1)`% of samples at an average relative abundance of `r d_gigas_summary[4]`% (median relative abundance = `r d_gigas_summary[3]`%).

Across all healthy controls available in `curatedMetagenomicData` (52 datasets), the presence of one or more of the sulfate / sulfite reducing bacteria were detected in `r round(sulf_red_RA_prev*100, 1)`% of samples.

# Prevalence of Cysteine Degrading Bacteria

[]

## Importing results from GutFunFind

Before reporting results similar to the sulfite and sulfate reducing bacteria for cysteine degrading bacteria, I used GutFunFind to query known cysteine degrading enzymes against the 4,644 representative UHGG genomes. Of these representative genomes, 252 generated a hit for 1 or more cysteine degrading enzyme. This generated a list of taxa containing likely homologues of these known enzymes. This taxonomic data has been imported from GutFunFind and is located in `results/from-GutFunFind/taxa_hits.txt`.

```{r}
taxa_hits <- read.csv("../results/from-GutFunFind/taxa_hits.txt", sep = "\t", header = FALSE)
taxa_hits <- as.vector(t(taxa_hits), mode = "character") # df -> vector
taxa_hits <- strsplit(taxa_hits, ";")

# extracting species names from taxa hits
taxa_hits_spp <- vector(mode = "character", length = length(taxa_hits)) # create empty caracter vector
for (i in 1:length(taxa_hits)) {
  current_spp <- gsub(pattern = "s__", replacement = "", taxa_hits[[i]][7])
  taxa_hits_spp[i] <- current_spp
}

# removing empty species names
taxa_hits_spp
length(taxa_hits_spp)
taxa_hits_spp <- taxa_hits_spp[taxa_hits_spp != ""]
taxa_hits_spp
length(taxa_hits_spp)

# removing Co-Abundance gene Groups (CAGs), GCAs (?), etc.
if (length(grep("CAG", taxa_hits_spp)) == 0) {
  # do nothing
} else if (length(grep("CAG", taxa_hits_spp)) > 0) {
  taxa_hits_spp <- taxa_hits_spp[-grep("CAG", taxa_hits_spp)]
}
taxa_hits_spp
length(taxa_hits_spp)
if (length(grep("GCA", taxa_hits_spp)) == 0) {
  # do nothing
} else if (length(grep("GCA", taxa_hits_spp)) > 0) {
  taxa_hits_spp <- taxa_hits_spp[-grep("GCA", taxa_hits_spp)]
}
taxa_hits_spp
length(taxa_hits_spp)
if (length(grep("sp[0-9]*", taxa_hits_spp)) == 0) {
  # do nothing
} else if (length(grep("sp[0-9]*", taxa_hits_spp)) > 0) {
  taxa_hits_spp <- taxa_hits_spp[-grep("sp[0-9]*", taxa_hits_spp)]
}
taxa_hits_spp
length(taxa_hits_spp)

# removing oddities from names like "_A", "_X", etc.
for (i in 1:length(taxa_hits_spp)) {
  current_name <- gsub("_.", "", taxa_hits_spp[i])
  current_name <- gsub(" ", "_", current_name)
  taxa_hits_spp[i] <- current_name
}
length(taxa_hits_spp)

# removing duplicates of species names
taxa_hits_spp <- unique(taxa_hits_spp)
length(taxa_hits_spp)
```

The above code takes an initial 493 taxa hits from `GutFunFind` and filters out names without species level resolition. A number of other modifications were also made to make the taxonomic name searchable against the identifiers present in `curatedMetagenomicData`. The variable `taxa_hits_spp` contains a final list of 254 Genus_species names with cysteine degrading potential.

## Prevalance of cysteine degrading bacteria in healthy controls

```{r}
indicies <- list(mode = "integer")
for (i in 1:length(taxa_hits_spp)) {
  indicies[[i]] <- grep(taxa_hits_spp[i], rownames(control_RA))
}
indicies <- unlist(indicies)
indicies <- unique(indicies) # sanity check, no repeat rows

# subsetting rel. abund. of cysteine deg bacteria
cys_deg_RA <- control_RA[indicies, ]
dim(cys_deg_RA)

# removing rows with __t in the taxa name... leading to double counting of RA
cys_deg_RA <- cys_deg_RA[-grep("t__", rownames(cys_deg_RA)), ]
dim(cys_deg_RA)

# computing colSums and calculating XX% prevalance of cys. deg. bac.
cys_deg_cs <- colSums(cys_deg_RA)
sum(cys_deg_cs > 0) / length(cys_deg_cs)
```

# Comparing abundance of Sulfi(a)te red. bac. and cys. deg. bac.

```{r}
# prepping data for plotting
sulf_red_RA <- t(sulf_red_RA)
sulf_red_cs <- colSums(sulf_red_RA)
sulf_red_summary <- summary(sulf_red_cs)
sulf_red_summary
cys_deg_summary <- summary(cys_deg_cs)
cys_deg_summary

df <- data.frame(sulf_red = sulf_red_cs,
                 cys_deg = cys_deg_cs)

# plotting abundances of sulf(ai)te red bac and cys deg bac
ggplot(stack(df), aes(x = ind, y = values)) +
  geom_boxplot(outlier.alpha = 0.12) + 
  stat_summary(fun.y = mean, geom = "point", shape = 4, size = 4) +
  scale_y_continuous(breaks = seq(0, 100, by = 10)) + 
  # geom_hline(yintercept = median(df$cys_deg), color = "red", linetype = "dashed") +
  ggtitle("Comparing relative abundances of two H2S producing pathways") +
  xlab("type of H2S producing bacteria") + ylab("Relative Abundance (%)") + 
  theme_linedraw()
ggsave("../figures/sulfRed_vs_cysDeg_abundances.png")
```

### Comparison of cysteine degrading and methionine degrading bacteria

```{r}
## load in pie chart file
tmp <- read.table("../results/from-GutFunFind/H2S_production.pichart.txt", skip = 7)
rownames(tmp) <- tmp$V1
tmp %>%
  select(V4, V5, V6, V7) -> tmp
colnames(tmp) <- c("both", "absent", "methionine", "cysteine")
set1 <- tmp %>% filter(methionine > 0) %>% rownames()
set2 <- tmp %>% filter(cysteine > 0) %>% rownames()
# set3 <- tmp %>% filter(both > 0) %>% rownames()

## venn diagram showing overlap between cys and met degrading functionality
venn.diagram(
  x = list(set1, set2),
  category.names = c("methionine" , "cysteine"),
  filename = "../figures/cys_met_overlap.png",
  output = TRUE,
          imagetype="png" ,
          height = 480,
          width = 480,
          resolution = 300,
          lwd = 1,
          col = c("blue", "red"),
          fill = c(alpha("blue",0.3), alpha('red',0.3)),
          cex = 0.5,
          fontfamily = "sans",
          cat.cex = 0.3,
          cat.default.pos = "outer",
          cat.pos = c(-27, 27),
          cat.dist = c(0.055, 0.055),
          cat.fontfamily = "sans",
          cat.col = c("blue", "red"),
)
```

```{r}
## comparison of cys / met degrading bac using upsetR
library(upsetR)


```

